src/transferslot.cpp

```diff
@@ -347,7 +347,100 @@ int64_t TransferSlot::macsmac(chunkmac_map* m)
+    // completed at the server end).  So the file's data is still complete in the cloud.
+    // Here we check if the MAC is one of those with a missing entry (or a few if the connection had multiple chunks)
+
+    // last 3 connections, up to 32MB (ie chunks) each, up to two completing after the one that delivered the ultoken
+    size_t end = transfer->chunkmacs.size();
+    size_t finalN = std::min<int>(32 * 3, end);
+
+    // first check for the most likely - a single connection gap (or two but completely consecutive making a single gap)
+    for (size_t countBack = 1; countBack <= finalN; ++countBack)
+    {
+        size_t start1 = end - countBack; 
+        for (size_t len1 = 1; len1 <= 64 && start1 + len1 <= end; ++len1)
+        {
+            if (transfer->metamac == macsmac_gaps(&transfer->chunkmacs, start1, start1 + len1, end, end))
+            {
+                LOG_warn << "Found mac gaps were at " << start1 << " " << len1 << " from " << end;
+                auto correctMac = macsmac(&transfer->chunkmacs);
+                transfer->currentmetamac = correctMac;
+                transfer->metamac = correctMac;
+                updateMacInKey(correctMac);
+                return true;
+            }
+        }
+    }
+
+    // now check for two separate pieces missing (much less likely)
+    // limit to checking up to 16Mb pieces wtih up to 8Mb between to avoid excessive CPU
+    // takes about 1 second on a fairly modest laptop for a 100Mb file (in a release build)
+    int caseschecked = 0;
+    finalN = std::min<int>(16 * 2 + 8, transfer->chunkmacs.size());
+    for (size_t start1 = end - finalN; start1 < end; ++start1)
+    {
+        for (size_t len1 = 1; len1 <= 16 && start1 + len1 <= end; ++len1)
+        {
+            for (size_t start2 = start1 + len1 + 1; start2 < transfer->chunkmacs.size(); ++start2)
+            {
+                for (size_t len2 = 1; len2 <= 16 && start2 + len2 <= end; ++len2)
+                {
+                    if (transfer->metamac == macsmac_gaps(&transfer->chunkmacs, start1, start1 + len1, start2, start2 + len2))
+                    {
+                        LOG_warn << "Found mac gaps were at " << start1 << " " << len1 << " " << start2 << " " << len2 << " from " << end;
+                        auto correctMac = macsmac(&transfer->chunkmacs);
+                        transfer->currentmetamac = correctMac;
+                        transfer->metamac = correctMac;
+                        updateMacInKey(correctMac);
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+    return false;
+}
+
+void TransferSlot::updateMacInKey(int64_t correctMac)
+{
+    // Update the Node's key to be correct (needs some API additions before enabling)
+    //for (file_list::iterator it = transfer->files.begin(); it != transfer->files.end(); it++)
+    //{
+    //    if ((*it)->hprivate && !(*it)->hforeign)
```

Comment:
Probably we want to allow sharees to rewrite keys too, not only the owner of the node.

Reply:
We keep this part to a different PR ;)